### ISSUE TITLE
Two potential issues in PPOTrainer: reward added to PAD token instead of EOS, and missing_eos_penalty may not apply

### ISSUE THREAD

**Issue by noncollapse** [huggingface/trl#3750]
### Reproduction

Hi, I’m currently reading through the implementation of PPOTrainer in the trl library and found two potential mismatches between the implementation and the intended behavior described in prior work on RLHF, particularly in *“The N+ Implementation Details of RLHF with PPO: A Case Study on TL;DR Summarization”*. 
## First: The position where the scores are added
In the current trl PPO training loop, the reward score is added to the token at the following index:
```python
actual_start = torch.arange(rewards.size(0), device=rewards.device)
actual_end = torch.where(sequence_lengths_p1 < rewards.size(1), sequence_lengths_p1, sequence_lengths)
rewards[[actual_start, actual_end]] += scores

```

where sequence_lengths_p1 is defined as:

```python
sequence_length = first_true_indices(postprocessed_response == processing_class.pad_token_id) - 1
sequence_lengths_p1 = sequence_lengths + 1
    ...
```
This means that in most cases, the reward score (scores) is added to the first **PAD token** in the sequence, rather than the **EOS token**.

However, as explained in the paper mentioned above, and also consistent with the description in the "*From r to Q∗: Your Language Model is Secretly a Q-Function*" paper (Rafailov et al., 2024), the reward score should ideally be assigned at the **EOS token**, since that is where the reward model is expected to backpropagate during training. If the EOS token is missing, the response may be penalized (e.g., with a score of -1), but when EOS is present, the reward should be anchored on it, not on the subsequent padding.

Assigning the score to the PAD token may introduce issues, such as if the PAD position is later masked (which is usually the case), the reward score may not actually contribute to training. 

## Second: The missing_eos_penalty
In the *Response Processing 3. Filter completion. Ensure that the sample contains stop_token_id* :
```python
contain_eos_token = torch.any(postprocessed_responses == self.processing_class.eos_token_id, dim=-1)
```
However, if we explicitly set:

```python
self.policy_model.generation_config.eos_token_id = self.stop_token_id = args.stop_token_id
```
Then generation will **correctly terminate at** the `stop_token_id` (e.g., `<|im_end|>`). Despite that, the above check still looks for the default `eos_token_id` (e.g., `<|endoftext|>`), so the condition fails even though the generation stopped as intended.

I’d like to ask whether this current behavior is intentional—perhaps for implementation simplicity or alignment with certain tokenization behaviors—or whether it might be an unintended bug that could affect training efficacy. Understanding the motivation behind this design choice would be very helpful.

I’d be happy to discuss this further.

Thanks for your great work on this library!

### System Info

None

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by behroozazarkhalili** [comment]
Thanks for the careful read. I traced the indexing on current `main` (PPO now lives at `trl/experimental/ppo/ppo_trainer.py`) and can confirm the factual part of your first point, with one clarification on where the reward lands.

Walking the indices for a response `[t0, t1, <eos>, <pad>, <pad>]`:
- `sequence_length = first_true_indices(response == pad_token_id) - 1` → `first_true_indices` returns the **first** pad index (`3`), so `sequence_length = 2` = the **EOS** position. ([`first_true_indices`](https://github.com/huggingface/trl/blob/main/trl/experimental/utils.py#L482) returns "the position of the first True … or row length if none".)
- `sequence_lengths_p1 = sequence_length + 1 = 3` = the **first PAD** position.
- `actual_end = where(sequence_lengths_p1 < size, sequence_lengths_p1, sequence_lengths)` → in the normal padded case this is `sequence_lengths_p1`, so `rewards[:, 3] += scores`.

So yes — in the common (padded) case the score is added at the **token immediately after EOS** (the first pad slot), not on the EOS token itself. The `where(...)` fallback to `sequence_lengths` only triggers when the response fills the whole length with no pad (so `seq+1` would be out of bounds).

The thing worth flagging before treating this as a bug: this is a faithful port of the "N+ Implementation Details of RLHF with PPO" reference implementation (the trainer cites [vwxyzjn's gist](https://github.com/huggingface/trl/blob/main/trl/experimental/ppo/ppo_trainer.py#L549)), where the reward is deliberately placed at the `seq+1` position and the value function is masked with `padding_mask_p1` (`response_idxs > sequence_lengths_p1`) at the same index — so the reward and the value bootstrap line up. Changing the reward to land on the EOS index would also need the `padding_mask_p1` / value alignment to move with it, otherwise the advantage at that position shifts. So this looks intentional-as-ported rather than an accidental off-by-one, though I agree it diverges from the "anchor on EOS" framing in the DPO-as-Q-function discussion.

@kashif — since PPO is now under `trl/experimental/`, is the `seq+1` reward placement (matching the original N+ reference impl) something you'd want to keep as-is, or revisit toward EOS-anchored rewards? Happy to put up a PR (reward placement + the corresponding `padding_mask`/value alignment + a test) if you'd like it changed.

**Comment by qgallouedec** [comment]
Hey @noncollapse, sorry for the very slow reply here — catching up on the backlog. Two different things, so let me take them separately.

Point 1 (reward at PAD vs EOS): this is intentional-as-ported, not an off-by-one. The score is added at `sequence_lengths_p1`-> the first PAD, i.e. the token right after EOS. 

https://github.com/huggingface/trl/blob/06b42c7e205437538d5fe3c39b53fdd911a46785/trl/experimental/ppo/ppo_trainer.py#L786-L788

the value function is masked with `padding_mask_p1` at the *same* index 
https://github.com/huggingface/trl/blob/06b42c7e205437538d5fe3c39b53fdd911a46785/trl/experimental/ppo/ppo_trainer.py#L776-L778

so the reward and the value bootstrap line up there. This follows the original "N+ Implementation Details of RLHF with PPO" reference impl; moving the reward onto the EOS index would also require moving the value/mask alignment with it, so we'd keep it as-is.

Point 2 (missing_eos_penalty): you're right, this one is a genuine inconsistency. Truncation and generation use `self.stop_token_id`, but the containment filter checks the tokenizer's default `eos_token_id` so a custom `stop_token_id` would wrongly trigger the penalty even when generation stopped correctly. It's niche (only when `stop_token_id != eos_token_id`) and PPO is now in maintenance mode. So it's unlikely that we will fix/review a fix. Consequently I'm going to close this.
